bindings/profiler.cc

```diff
@@ -85,73 +82,142 @@ NAN_METHOD(StopSamplingHeapProfiler) {
   }
-  js_node->Set(Nan::New<String>("children").ToLocalChecked(),
-    children);
+
+  js_node->Set(Nan::New<String>("children").ToLocalChecked(), children);
   return js_node;
 }
 
-Local<Value> TranslateTimeProfile(const CpuProfile* profile) {
+Local<Value> TranslateTimeProfile(const CpuProfile* profile, bool hasLines) {
   Local<Object> js_profile = Nan::New<Object>();
   js_profile->Set(Nan::New<String>("title").ToLocalChecked(),
-    profile->GetTitle());
-  js_profile->Set(Nan::New<String>("topDownRoot").ToLocalChecked(),
-    TranslateTimeProfileNode(profile->GetTopDownRoot()));
+                  profile->GetTitle());
+  js_profile->Set(
+      Nan::New<String>("topDownRoot").ToLocalChecked(),
+      TranslateTimeProfileNode(profile->GetTopDownRoot(), hasLines));
   js_profile->Set(Nan::New<String>("startTime").ToLocalChecked(),
-    Nan::New<Number>(profile->GetStartTime()));
+                  Nan::New<Number>(profile->GetStartTime()));
   js_profile->Set(Nan::New<String>("endTime").ToLocalChecked(),
-    Nan::New<Number>(profile->GetEndTime()));
+                  Nan::New<Number>(profile->GetEndTime()));
   return js_profile;
 }
 
 NAN_METHOD(StartProfiling) {
-  Local<String> name = info[0].As<String>();
+  if (info.Length() != 2) {
+    return Nan::ThrowTypeError("StartProfling must have two arguments.");
+  }
+  if (!info[0]->IsString()) {
+    return Nan::ThrowTypeError("First argument type must be a string.");
+  }
+  if (!info[1]->IsBoolean()) {
+    return Nan::ThrowTypeError("First argument type must be a boolean.");
+  }
+
+  Local<String> name =
+      Nan::MaybeLocal<String>(info[0].As<String>()).ToLocalChecked();
+  bool includeLineInfo =
+      Nan::MaybeLocal<Boolean>(info[1].As<Boolean>()).ToLocalChecked()->Value();
 
-  // Sample counts and timestamps are not used, so we do not need to record
-  // samples.
+// Sample counts and timestamps are not used, so we do not need to record
+// samples.
+#if NODE_MODULE_VERSION > NODE_10_0_MODULE_VERSION
+  if (includeLineInfo) {
+    cpuProfiler->StartProfiling(name, CpuProfilingMode::kCallerLineNumbers,
+                                false);
+  } else {
+    cpuProfiler->StartProfiling(name, false);
+  }
+#else
   cpuProfiler->StartProfiling(name, false);
+#endif
 }
 
 NAN_METHOD(StopProfiling) {
-  Local<String> name = info[0].As<String>();
-  CpuProfile* profile =
-    cpuProfiler->StopProfiling(name);
-  Local<Value> translated_profile = TranslateTimeProfile(profile);
+  if (info.Length() != 2) {
+    return Nan::ThrowTypeError("StopProfling must have two arguments.");
+  }
+  if (!info[0]->IsString()) {
+    return Nan::ThrowTypeError("First argument type must be a string.");
+  }
+  if (!info[1]->IsBoolean()) {
+    return Nan::ThrowTypeError("First argument type must be a boolean.");
+  }
+  Local<String> name =
+      Nan::MaybeLocal<String>(info[0].As<String>()).ToLocalChecked();
+  bool includedLineInfo =
```

Comment:
Is the expectation that the caller must pass the same value for `includedLineInfo` as what they passed to `StartProfiling`. What happens when they pass in the wrong value here?

It might be better to keep this state on the C++ side and not require the caller to pass the same value to `Start` and `StopProfiling`.

Reply:
There is the expectation that the same values are passed in. 

If, for example, `includeLineInfo=true` for `StartProfiling`, and `false` for `StopProfiling`, then LineTick information would not be included in the returned profile, so it wouldn't be possible to get accurate line numbers for leaf nodes. 

I guess this variable could be renamed to "includeLineTicks" in stop profiling. 

I'm not sure about storing this as state, since we'd have to basically store a map of name used to start profile to that boolean value, and I'd be worried about not deleting things from the array properly.

Reply:
Do we have test cases that exercise incorrect usage of the API (and ensure that the failure case isn't anything bad). That is, a test case that passes in mismatched value for `includeLineInfo` between start and stop.